File: tinytorch/activations.py
"""Activations — element-wise nonlinear activation functions.

All functions accept a Tensor and return a new Tensor.
No learnable parameters. Forward-only in E02;
backward implementations will be added in E06.
"""

from __future__ import annotations

from tinytorch.tensor import Tensor


def relu(x: Tensor) -> Tensor:
    """ReLU activation: f(x) = max(0, x)."""
    raise NotImplementedError("TODO: E02")


def sigmoid(x: Tensor) -> Tensor:
    """Sigmoid activation: f(x) = 1 / (1 + exp(-x))."""
    raise NotImplementedError("TODO: E02")


def tanh(x: Tensor) -> Tensor:
    """Tanh activation: f(x) = (exp(x) - exp(-x)) / (exp(x) + exp(-x))."""
    raise NotImplementedError("TODO: E02")


def gelu(x: Tensor) -> Tensor:
    """GELU activation (tanh approximation):
    f(x) = x * 0.5 * (1 + tanh(sqrt(2/pi) * (x + 0.044715 * x^3))).
    """
    raise NotImplementedError("TODO: E02")


def softmax(x: Tensor, axis: int) -> Tensor:
    """Softmax activation: f(x_i) = exp(x_i - max) / sum(exp(x_j - max)).

    Numerically stable: subtracts the max before computing exp.
    """
    raise NotImplementedError("TODO: E02")


# ================================================================
# E06 — More Backward Ops
# ================================================================
# When x.requires_grad is True, relu(), sigmoid(), tanh(), and gelu()
# should create the corresponding Backward function (ReLUBackward,
# SigmoidBackward, TanhBackward, GELUBackward), call fn.forward(),
# and set result.requires_grad = True, result.grad_fn = fn.
# softmax does NOT need backward here (cross_entropy uses fusion).

File: src/main/java/dev/tensorhero/tinytorch/Activations.java
package dev.tensorhero.tinytorch;

import dev.tensorhero.tinynum.NDArray;

/**
 * Activations — element-wise nonlinear activation functions.
 *
 * <p>All methods are static, accept a {@link Tensor}, and return a new
 * {@link Tensor}. No learnable parameters. Forward-only in E02;
 * backward implementations will be added in E06.</p>
 */
public class Activations {

    // ================================================================
    // E02 — Activations
    // ================================================================

    /**
     * ReLU activation: f(x) = max(0, x).
     *
     * @param x the input tensor
     * @return a new Tensor with ReLU applied element-wise
     */
    public static Tensor relu(Tensor x) {
        throw new UnsupportedOperationException("TODO: E02");
    }

    /**
     * Sigmoid activation: f(x) = 1 / (1 + exp(-x)).
     *
     * @param x the input tensor
     * @return a new Tensor with sigmoid applied element-wise
     */
    public static Tensor sigmoid(Tensor x) {
        throw new UnsupportedOperationException("TODO: E02");
    }

    /**
     * Tanh activation: f(x) = (exp(x) - exp(-x)) / (exp(x) + exp(-x)).
     *
     * @param x the input tensor
     * @return a new Tensor with tanh applied element-wise
     */
    public static Tensor tanh(Tensor x) {
        throw new UnsupportedOperationException("TODO: E02");
    }

    /**
     * GELU activation (tanh approximation):
     * f(x) = x * 0.5 * (1 + tanh(sqrt(2/pi) * (x + 0.044715 * x^3))).
     *
     * @param x the input tensor
     * @return a new Tensor with GELU applied element-wise
     */
    public static Tensor gelu(Tensor x) {
        throw new UnsupportedOperationException("TODO: E02");
    }

    /**
     * Softmax activation: f(x_i) = exp(x_i - max) / sum(exp(x_j - max)).
     *
     * <p>Numerically stable: subtracts the max before computing exp
     * to prevent overflow.</p>
     *
     * @param x    the input tensor
     * @param axis the axis along which to compute softmax
     * @return a new Tensor with softmax applied along the given axis
     */
    public static Tensor softmax(Tensor x, int axis) {
        throw new UnsupportedOperationException("TODO: E02");
    }

    // ================================================================
    // E06 — More Backward Ops
    // ================================================================
    // When x.requiresGrad is true, relu(), sigmoid(), tanh(), and gelu()
    // should create the corresponding Backward function (ReLUBackward,
    // SigmoidBackward, TanhBackward, GELUBackward), call fn.forward(),
    // and set result.requiresGrad = true, result.gradFn = fn.
    // softmax does NOT need backward here (CrossEntropy uses fusion).
}
